File: pg/BO/ssap-partner-portal-bo-frontend/src/models/Master.ts
import {
  IGetAlertSettingResponse,
  IMasterListItem,
  IGetMastersRequest,
  IGetMastersResponse,
  IGetMasterByIdRequest,
  IGetMasterByIdResponse,
  ICreateMasterResponse,
  IUpdateMasterRequestParams,
  IUpdateMasterResponse,
  IUpdateMasterRequest,
  ICreateMasterRequest,
  IGetSurveyMastersRequest,
  IGetSurveyMastersResponse,
  IServiceTypeItems,
  IGetServiceTypesResponse,
  ISupporterOrganizationItems,
  IGetSupporterOrganizationsResponse,
  ISupporterOrganizationsResponse,
  IGetSurveyMastersByIdResponse,
  ICopySurveyMasterByIdResponse,
  IPatchSurveyMasterRevisionByIdResponse,
  ISurveyMasterChoiceGroupItem,
  ISurveyMasterChoiceItem,
  ISurveyMasterQuestionFlowItem,
  ISurveyMasterQuestionItem,
  ICreateSurveyMastersRequest,
  ICreateSurveyMastersResponse,
  IGetSurveyMasterByIdAndRevisionResponse,
  IUpdateSurveyMasterLatestByIdRequest,
  IUpdateSurveyMasterLatestByIdResponse,
  IGetBatchControlByIdResponse,
  IUpdateAlertSettingsResponse,
  IUpdateAlertSettingsRequest,
  IUpdateSurveyMasterDraftByIdRequest,
  IAttributes,
  IAlertSettingAttributes,
  IDisplaySetting,
  IFactorSetting,
  IDirectSupportManHour,
  IPreSupportManHour,
  GetSurveyMastersIsLatest,
  ENUM_GET_SURVEY_MASTERS_IS_LATEST,
} from '@/types/Master'
import { Api } from '~/plugins/api'

const $api = new Api()

export class MasterListItem implements IMasterListItem {
  public id = ''
  public dataType = ''
  public name = ''
  public value = ''
  public use = false
  public maintenance = false
  public createAt = ''
  public updateAt = ''
}

export class GetMastersRequest implements IGetMastersRequest {
  public dataType = 'all'
  public limit = 20
  public offsetPage = 1
}
export class GetMastersResponse implements IGetMastersResponse {
  public offsetPage = 1
  public total = 0
  public masters = []
}

export async function GetMasters(request: GetMastersRequest) {
  return await $api.get<GetMastersResponse>(`/masters`, request)
}

export class GetMasterByIdRequest implements IGetMasterByIdRequest {
  public id = ''
}

export class Attributes implements IAttributes {
  public info1 = ''
  public info2 = ''
  public info3 = ''
  public info4 = ''
  public info5 = ''
}

export class GetMasterByIdResponse implements IGetMasterByIdResponse {
  public id = ''
  public dataType = ''
  public name = ''
  public value = ''
  public attributes = new Attributes()
  public use = true
  public createAt = ''
  public updateAt = ''
  public version = 0
}

export function GetMasterById(id: string) {
  return $api.get<GetMasterByIdResponse>(`/masters/${id}`)
}

export class CreateMasterRequest implements ICreateMasterRequest {
  public dataType = ''
  public name = ''
  public value = ''
  public attributes = new Attributes()
  public use = false
}

export class CreateMasterResponse implements ICreateMasterResponse {
  public id = ''
  public dataType = ''
  public name = ''
  public value = ''
  public attributes = new Attributes()
  public use = false
}

export async function CreateMaster(request: CreateMasterRequest) {
  return await $api.post<CreateMasterResponse>(`/masters`, request)
}

export class UpdateMasterRequestParams implements IUpdateMasterRequestParams {
  public id = ''
  public version = 0
}

export class UpdateMasterRequest implements IUpdateMasterRequest {
  public name = ''
  public value = ''
  public attributes = new Attributes()
  public order = 1
  public use = true
  public maintenance = true
}

export class UpdateMasterResponse implements IUpdateMasterResponse {
  public dataType = ''
  public name = ''
  public value = ''
  public attributes = new Attributes()
  public use = false
}

export async function UpdateMasterById(
  id: string,
  version: number,
  request: UpdateMasterRequest
) {
  return await $api.put<UpdateMasterResponse>(
    `/masters/${id}?version=${version}`,
    request
  )
}
export class GetSurveyMastersRequest implements IGetSurveyMastersRequest {
  name = ''
  latest = true
}

export class SurveyMasterListItem {
  public id = ''
  public revision = 0
  public name = ''
  public type = ''
  public timing = ''
  public initSendDaySetting = 0
  public initAnswerLimitDaySetting = 0
  public isDisclosure = false
  public isLatest: GetSurveyMastersIsLatest =
    ENUM_GET_SURVEY_MASTERS_IS_LATEST.FALSE

  public memo = ''
  public status = ''
  public questionCount = 0
  public updateAt = ''
}
export class GetSurveyMastersResponse implements IGetSurveyMastersResponse {
  total = 0
  masters: SurveyMasterListItem[] = []
}

export async function GetSurveyMasters(request: GetSurveyMastersRequest) {
  return await $api.get<GetSurveyMastersResponse>(`/survey-masters`, request)
}

export class ServiceTypeItems implements IServiceTypeItems {
  public id = ''
  public name = ''
}
export class GetServiceTypesResponse implements IGetServiceTypesResponse {
  public serviceTypes: IServiceTypeItems[] = []
}

export async function GetServiceTypes() {
  return await $api.get<GetServiceTypesResponse>(`/masters/service-types`)
}

export class SupporterOrganizationItems implements ISupporterOrganizationItems {
  public id = ''
  public name = ''
  public shortName? = ''
}

export class GetSupporterOrganizationsResponse
  implements IGetSupporterOrganizationsResponse
{
  public supporterOrganizations: ISupporterOrganizationItems[] = []
}

export async function GetSupporterOrganizations() {
  return await $api.get<GetSupporterOrganizationsResponse>(
    `/masters/supporter-organizations`
  )
}

export class SupporterOrganizationsResponse
  implements ISupporterOrganizationsResponse
{
  public id = ''
  public name = ''
  public shortName = ''
}
export class GetSurveyMastersByIdResponse
  implements IGetSurveyMastersByIdResponse
{
  public total = 0
  public masters: SurveyMasterListItem[] = []
}

export class CopySurveyMasterByIdResponse
  implements ICopySurveyMasterByIdResponse
{
  public status = ''
  public revision = 0
  public memo = ''
  public questionsCont = 0
  public updateAt = ''
}

export class PatchSurveyMasterRevisionByIdResponse
  implements IPatchSurveyMasterRevisionByIdResponse
{
  public status = ''
  public revision = 0
  public memo = ''
  public questionsCont = 0
  public updateAt = ''
}

export class SurveyMasterChoiceGroupItem
  implements ISurveyMasterChoiceGroupItem
{
  public id = ''
  public title = ''
  public disabled = false
  public isNew = false
}

export class SurveyMasterChoiceItem implements ISurveyMasterChoiceItem {
  public description = ''
  public group: SurveyMasterChoiceGroupItem[] = []
  public isNew = false
}

export class SurveyMasterQuestionFlowItem
  implements ISurveyMasterQuestionFlowItem
{
  public id = ''
  public conditionId = ''
  public conditionChoiceIds: string[] = []
}

export class SurveyMasterQuestionItem implements ISurveyMasterQuestionItem {
  public id = ''
  public required = false
  public description = ''
  public format = ''
  public summaryType = ''
  public choices: SurveyMasterChoiceItem[] = []
  public otherDescription = ''
  public disabled = false
  public isNew = false
}

export class CreateSurveyMastersRequest implements ICreateSurveyMastersRequest {
  public name = ''
  public type = ''
  public timing = ''
  public initSendDaySetting = 0
  public initAnswerLimitDaySetting = 0
  public isDisclosure = false
  public questions: SurveyMasterQuestionItem[] = []
  public questionFlow: SurveyMasterQuestionFlowItem[] = []
  public memo = ''
}

export class CreateSurveyMastersResponse
  implements ICreateSurveyMastersResponse
{
  public id = ''
  public revision = 0
  public name = ''
  public type = ''
  public timing = ''
  public initSendDaySetting = 0
  public initAnswerLimitDaySetting = 0
  public isDisclosure = false
  public questions: SurveyMasterQuestionItem[] = []
  public questionFlow: SurveyMasterQuestionFlowItem[] = []
  public isLatest = 0
  public memo = ''
  public createId = ''
  public createUserName = ''
  public createAt = ''
  public version = 0
}

export class GetSurveyMasterByIdAndRevisionResponse
  implements IGetSurveyMasterByIdAndRevisionResponse
{
  public id = ''
  public revision = 0
  public name = ''
  public type = ''
  public timing = ''
  public initSendDaySetting = 0
  public initAnswerLimitDaySetting = 0
  public isDisclosure = false
  public questions: SurveyMasterQuestionItem[] = []
  public questionFlow: SurveyMasterQuestionFlowItem[] = []
  public isLatest = 0
  public memo = ''
  public createId = ''
  public createUserName = ''
  public createAt = ''
  public updateId = ''
  public updateUserName = ''
  public updateAt = ''
  public version = 0
}

export class UpdateSurveyMasterLatestByIdRequest
  implements IUpdateSurveyMasterLatestByIdRequest
{
  public name = ''
  public type = ''
  public timing = ''
  public initSendDaySetting = 0
  public initAnswerLimitDaySetting = 0
  public isDisclosure = false
  public memo = ''
}

export class UpdateSurveyMasterLatestByIdResponse
  implements IUpdateSurveyMasterLatestByIdResponse
{
  public id = ''
  public revision = 0
  public name = ''
  public type = ''
  public timing = ''
  public initSendDaySetting = 0
  public initAnswerLimitDaySetting = 0
  public isDisclosure = false
  public isLatest = 0
  public memo = ''
  public updateId = ''
  public updateUserName = ''
  public updateAt = ''
  public version = 0
}

export async function GetSurveyMastersById(id: string) {
  return await $api.get<GetSurveyMastersByIdResponse>(`/survey-masters/${id}`)
}

export async function CopySurveyMasterById(id: string, originRevision: number) {
  return await $api.put(
    `/survey-masters/${id}?originRevision=${originRevision}`
  )
}

export async function PatchSurveyMasterRevisionById(
  id: string,
  version: number
) {
  return await $api.patch(`/survey-masters/${id}?version=${version}`)
}

export async function CreateSurveyMasters(request: CreateSurveyMastersRequest) {
  return await $api.post<CreateSurveyMastersResponse>(
    `/survey-masters`,
    request
  )
}

export async function GetSurveyMasterByIdAndRevision(
  id: string,
  revision: number
) {
  return await $api.get<GetSurveyMasterByIdAndRevisionResponse>(
    `/survey-masters/${id}/${revision}`
  )
}

export async function UpdateSurveyMasterLatestById(
  id: string,
  version: number,
  request: UpdateSurveyMasterLatestByIdRequest
) {
  return await $api.put<UpdateSurveyMasterLatestByIdResponse>(
    `/survey-masters/${id}/latest?version=${version}`,
    request
  )
}

export class GetBatchControlByIdResponse
  implements IGetBatchControlByIdResponse
{
  public batchEndAt = ''
}

export async function GetBatchControlById(id: string) {
  return await $api.get<GetBatchControlByIdResponse>(
    `/masters/batch-control/${id}`
  )
}

export class FactorSetting implements IFactorSetting {
  serviceTypeId = ''
  directSupportManHour = 0
  directAndPreSupportManHour = 0
  preSupportManHour = 0
  hourlyManHourPrice = 0
  monthlyProfit = 0
}
export class DirectSupportManHour implements IDirectSupportManHour {
  summaryDisplayProject = true
  summaryOverAlert = true
  summaryProspectAlert = true
  thisMonthDisplayProject = true
  thisMonthOverAlert = true
  thisMonthProspectAlert = true
}

export class PreSupportManHour implements IPreSupportManHour {
  summaryDisplayProject = true
  summaryOverAlert = true
  summaryProspectAlert = true
  thisMonthDisplayProject = true
  thisMonthOverAlert = true
  thisMonthProspectAlert = true
}
export class DisplaySetting implements IDisplaySetting {
  directSupportManHour: DirectSupportManHour = new DirectSupportManHour()
  preSupportManHour: PreSupportManHour = new PreSupportManHour()
}
export class AlertSettingAttributes implements IAlertSettingAttributes {
  factorSetting: FactorSetting[] = []
  displaySetting: IDisplaySetting = new DisplaySetting()
}

export class GetAlertSettingsResponse implements IGetAlertSettingResponse {
  id = ''
  name = ''
  attributes: AlertSettingAttributes = new AlertSettingAttributes()
  createId = ''
  createUserName = ''
  createAt = ''
  updateId = ''
  updateUserName = ''
  updateAt = ''
  version = 0
}

export async function GetAlertSettings() {
  return await $api.get<GetAlertSettingsResponse>(`/masters/alert-setting`)
}

export async function UpdateAlertSettings(
  version: number,
  request: UpdateAlertSettingsRequest
) {
  return await $api.put<UpdateAlertSettingsResponse>(
    `/masters/alert-setting?version=${version}`,
    request
  )
}

export class UpdateAlertSettingsRequest implements IUpdateAlertSettingsRequest {
  factorSetting = new FactorSetting()
  displaySetting? = new DisplaySetting()
}

export class UpdateAlertSettingsResponse
  implements IUpdateAlertSettingsResponse
{
  id = ''
  name = ''
  attributes = new AlertSettingAttributes()
  createId = ''
  createUserName = ''
  createAt = ''
  updateId = ''
  updateUserName = ''
  updateAt = ''
  version = 0
}

export class UpdateSurveyMasterDraftByIdRequest
  implements IUpdateSurveyMasterDraftByIdRequest
{
  public name = ''
  public timing = ''
  public initSendDaySetting = 0
  public initAnswerLimitDaySetting = 0
  public isDisclosure = false
  public questions: SurveyMasterQuestionItem[] = []
  public questionFlow: SurveyMasterQuestionFlowItem[] = []
  public memo = ''
}

export async function UpdateSurveyMasterDraftById(
  id: string,
  version: number,
  request: UpdateSurveyMasterDraftByIdRequest
) {
  return await $api.put(
    `/survey-masters/${id}/draft?version=${version}`,
    request
  )
}

File: pg/iret/BO/ssap-partner-portal-bo-backend/app/schemas/master.py
from datetime import datetime
from typing import Dict, List, Union

from fastapi import Query
from pydantic import BaseModel, Field, validator

from app.resources.const import (
    DefaultPageItemCount,
    MasterDataType,
    QueryParameterForMasterDataType,
)
from app.schemas.base import CustomBaseModel


class ReserveInformation(BaseModel):
    info1: str = Field(None, title="予備情報1", example="string")
    info2: str = Field(None, title="予備情報2", example="string")
    info3: str = Field(None, title="予備情報3", example="string")
    info4: str = Field(None, title="予備情報4", example="string")
    info5: str = Field(None, title="予備情報5", example="string")

    class Config:
        orm_mode = True


class MasterItem(CustomBaseModel):
    """マスター共通クラス"""

    id: str = Field(..., title="データID", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d")
    data_type: str = Field(..., title="種別", example="master_supporter_organization")
    name: str = Field(..., title="名称", example="Ideation Service Team")
    value: str = Field(..., title="値", example="IST")
    attributes: ReserveInformation = Field(None, title="予備情報")
    order: int = Field(..., title="登録順", example=1)
    use: bool = Field(..., title="利用フラグ", example=True)
    create_id: str = Field(
        None, title="作成者ID", example="fa5c9f96-af22-bbc6-ac4e-cbcb25ad18e7"
    )
    create_at: datetime = Field(
        None, title="作成日時", example="2020-10-23T03:21:39.356+09:00"
    )
    update_id: str = Field(
        None, title="作成者ID", example="fa5c9f96-af22-bbc6-ac4e-cbcb25ad18e7"
    )
    update_at: datetime = Field(
        None, title="更新日時", example="2020-10-23T03:21:39.356+09:00"
    )
    version: int = Field(None, title="楽観ロックバージョン", example=1)

    class Config:
        orm_mode = True


class CreateMasterRespose(MasterItem):
    """マスター作成レスポンスクラス"""

    pass


class CreateMasterRequest(CustomBaseModel):
    """マスター作成リクエストクラス"""

    data_type: str = Field(..., title="種別", example="master_supporter_organization")
    name: str = Field(..., title="名称", example="Ideation Service Team")
    value: str = Field(..., title="値", example="IST")
    attributes: ReserveInformation = Field(None, title="予備情報")
    use: bool = Field(..., title="利用フラグ", example=True)

    @validator("data_type")
    def check_customer_category(cls, v: str) -> Union[str, ValueError]:
        if v is not None:
            data_type = [c.value for _, c in MasterDataType.__members__.items()]
            if v not in data_type:
                raise ValueError("Category should be either of " + ", ".join(data_type))

        return v


class GetMastersMasterQuery(CustomBaseModel):
    """マスター一覧取得マスタークエリクラス"""

    data_type: QueryParameterForMasterDataType = Query(
        QueryParameterForMasterDataType.ALL,
        title="データ区分（支援者組織情報、サービス種別、工数アラート設定、バッチ処理制御）",
    )
    limit: int = Query(
        DefaultPageItemCount.limit,
        title="最大取得件数",
        example=20,
    )
    offset_page: int = Query(1, title="リストの中で何ページ目を取得するか", example=1)


class GetMastersMasterResponse(CustomBaseModel):
    """マスター一覧取得マスターレスポンスクラス"""

    id: str = Field(..., title="データID", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d")
    data_type: str = Field(..., title="種別", example="master_supporter_organization")
    name: str = Field(..., title="名称", example="Ideation Service Team")
    value: str = Field(..., title="値", example="IST")
    order: int = Field(..., title="登録順", example=1)
    use: bool = Field(..., title="利用フラグ", example=True)
    create_at: datetime = Field(
        ..., title="作成日時", example="2020-10-23T03:21:39.356+09:00"
    )
    update_at: datetime = Field(
        ..., title="更新日時", example="2020-10-23T03:21:39.356+09:00"
    )


class GetMastersResponse(CustomBaseModel):
    """マスター一覧取得レスポンスクラス"""

    offset_page: int = Field(None, title="リストの総数内で何ページ目か")
    total: int = Field(..., title="件数", ge=0)
    masters: List[GetMastersMasterResponse]


class GetMasterByIdResponse(MasterItem):
    """マスター詳細取得レスポンスクラス"""

    pass


class UpdateMasterByIdQuery(CustomBaseModel):
    version: int = Query(..., ge=1)


class UpdateMasterByIdRequest(CustomBaseModel):
    name: str = Field(..., title="名称", example="Ideation Service Team")
    value: str = Field(..., title="値", example="IST")
    attributes: ReserveInformation = Field(None, title="予備情報")
    use: bool = Field(..., title="利用フラグ", example=True)


class UpdateMasterByIdResponse(MasterItem):
    pass


class GetServiceTypesQuery(CustomBaseModel):
    """サービス種別一覧取得クエリクラス"""

    disabled: bool = Query(
        False,
        title="無効データ含む",
    )


class ServiceTypes(CustomBaseModel):
    """サービス種別(id, name)"""

    id: str = Field(..., title="id", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d")
    name: str = Field(..., title="名称", example="アイデア可視化")


class GetServiceTypesResponse(CustomBaseModel):
    service_types: List[ServiceTypes] = Field(..., title="サービス種別")


class GetBatchControlResponse(CustomBaseModel):
    """バッチ処理最終完了日時取得レスポンスクラス"""

    batch_end_at: datetime = Field(
        ..., title="処理終了日時（JST）", example="2020-10-23T03:21:39.356+09:00"
    )


class FactorSetting(CustomBaseModel):
    """アラート設定クラス"""

    service_type_id: str = Field(..., title="サービス種別ID")
    direct_support_man_hour: float = Field(..., title="対面支援工数係数（%）", ge=0, le=999.99)
    direct_and_pre_support_man_hour: float = Field(
        ..., title="対面仕込支援工数係数（%）", ge=0, le=999
    )
    pre_support_man_hour: int = Field(None, title="仕込み工数係数（%）")
    hourly_man_hour_price: int = Field(None, title="契約1時間当たりの標準的な時間単価（万円）")
    monthly_profit: int = Field(None, title="月の粗利目標（万円）")


class AlertSettingAttribute(CustomBaseModel):
    """アラート設定属性クラス"""

    factor_setting: List[FactorSetting] = Field(..., title="係数設定")
    # 実装時点では不要なため型の定義のみ
    display_setting: Dict = Field(None, title="アラート表示設定")


class GetAlertSettingResponse(CustomBaseModel):
    """工数アラート設定一覧取得レスポンスクラス"""

    id: str = Field(
        ..., title="工数アラート設定ID", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d"
    )
    name: str = Field(..., title="工数アラート設定名", example="支援工数アラート設定")
    attributes: AlertSettingAttribute = Field(..., title="制御情報")
    create_id: str = Field(
        ..., title="作成者ID", example="fa5c9f96-af22-bbc6-ac4e-cbcb25ad18e7"
    )
    create_at: datetime = Field(
        ..., title="作成日時", example="2020-10-23T03:21:39.356+09:00"
    )
    update_id: str = Field(
        ..., title="作成者ID", example="fa5c9f96-af22-bbc6-ac4e-cbcb25ad18e7"
    )
    update_at: datetime = Field(
        ..., title="更新日時", example="2020-10-23T03:21:39.356+09:00"
    )
    version: int = Field(..., title="楽観ロックバージョン", example=1)


class GetSupporterOrganizationsQuery(CustomBaseModel):
    """支援者組織一覧取得クエリクラス"""

    disabled: bool = Query(
        False,
        title="無効データ含む",
    )


class GetSupporterOrganizationsResponseItems(CustomBaseModel):
    id: str = Field(..., title="id", example="89cbe2ed-f44c-4a1c-9408-c67b0ca2270d")
    name: str = Field(..., title="名称", example="Ideation Service Team​")
    short_name: str = Field(..., title="値", example="IST")


class GetSupporterOrganizationsResponse(CustomBaseModel):
    supporter_organizations: List[GetSupporterOrganizationsResponseItems] = Field(
        ..., title="支援者組織"
    )
